File: python3/flatten_binary_tree_to_linked_list.py
from .model import TreeNode

# Space : O(n)
# Time  : O(n)

class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        
        def preorder(root):
            res = []
            
            if root:
                res.append(root.val)
                res += preorder(root.left)
                res += preorder(root.right)
                
            return res
        
        temp = preorder(root)
        n = len(temp)
        p = root

        for i in range(n):
            if i == 0:
                p.val = temp[i]
            else:
                p.right = TreeNode(temp[i])
                p.left = None
                p = p.right

File: golang/flatten_binary_tree_to_linked_list.go
package golang

// Space : O(n)
// Time  : O(n)

import {
	"golang/model"
}

func getTreePreorder(root *model.TreeNode) []int {
	res := []int{}

	if root != nil {
		res = append(res, root.Val)
		if root.Left != nil {
			res = append(res, getTreePreorder(root.Left)...)
		}
		if root.Right != nil {
			res = append(res, getTreePreorder(root.Right)...)
		}
	}

	return res
}

func flatten(root *model.TreeNode) {
	preorder := getTreePreorder(root)

	p := root
	for i, val := range preorder {
		p.Val = val
		p.Left = nil
		if i != len(preorder)-1 {
			p.Right = &model.TreeNode{}
		}

		p = p.Right
	}

}
